components/shared/webrender/viewport_description.rs

```diff
@@ -0,0 +1,183 @@
+                },
+                "initial-scale" => {
+                    if let Some(zoom) = Self::parse_viewport_value_as_zoom(value) {
+                        description.initial_scale = zoom;
+                    }
+                },
+                "minimum-scale" => {
+                    if let Some(zoom) = Self::parse_viewport_value_as_zoom(value) {
+                        description.minimum_scale = zoom;
+                    }
+                },
+                "maximum-scale" => {
+                    if let Some(zoom) = Self::parse_viewport_value_as_zoom(value) {
+                        description.maximum_scale = zoom;
+                    }
+                },
+                "user-scalable" => {
+                    if let Some(user_zoom) = Self::parse_viewport_value_as_user_zoom(value) {
+                        description.user_scalable = user_zoom;
+                    }
+                },
+                _ => (),
+            }
+        }
+        description
+    }
+
+    /// Parses a viewport length value.
+    fn parse_viewport_value_as_length(value: &str) -> ViewportLength {
+        match value.to_lowercase().as_str() {
+            "device-width" => ViewportLength::DeviceWidth,
+            "device-height" => ViewportLength::DeviceHeight,
+            _ => value
+                .parse::<f32>()
+                .ok()
+                .filter(|&n| n >= 0.0)
+                .map(ViewportLength::Fixed)
+                .unwrap_or(ViewportLength::Auto),
+        }
+    }
+
+    fn parse_viewport_value_as_zoom(value: &str) -> Option<Scale<f32>> {
+        value
+            .to_lowercase()
+            .as_str()
+            .parse::<f32>()
+            .ok()
+            .filter(|&n| (0.0..=10.0).contains(&n))
+            .map(Scale::new)
+    }
+
+    /// Parses a viewport user zoom value.
+    fn parse_viewport_value_as_user_zoom(value: &str) -> Option<UserZoom> {
+        Some(match value.to_lowercase().as_str() {
+            "yes" => UserZoom::Zoom,
+            "no" => UserZoom::Fixed,
+            _ => match value.parse::<f32>() {
+                Ok(n) if n >= 1.0 => UserZoom::Zoom,
+                Ok(_) => UserZoom::Fixed,
+                _ => return None,
+            },
+        })
+    }
+}
+
+impl FromStr for ViewportDescription {
+    type Err = ();
+    fn from_str(string: &str) -> Result<Self, Self::Err> {
+        if string.is_empty() {
+            return Err(());
+        }
+
+        // Parse key-value pairs from the content string
+        let parsed_values = string
+            .split([',', ';'])
+            .filter_map(|pair| {
+                let mut parts = pair.split('=').map(str::trim);
+                if let (Some(key), Some(value)) = (parts.next(), parts.next()) {
+                    Some((key.to_string(), value.to_string()))
+                } else {
+                    None
+                }
+            })
+            .collect::<HashMap<String, String>>();
```

Comment:
I would link to the part of the specification where viewport parsing is specified. It doesn't look like this is the same algorithm: https://drafts.csswg.org/css-viewport/#parsing-algorithm